File: aoc-2023/src/test/java/com/capital7software/aoc/aoc2023/days/Day12Test.java
package com.capital7software.aoc.aoc2023.days;

import com.capital7software.aoc.lib.AdventOfCodeTestBase;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import static org.junit.jupiter.api.Assertions.assertEquals;

class Day12Test extends AdventOfCodeTestBase {
    private static final Logger LOGGER = LoggerFactory.getLogger(Day12Test.class);

    @BeforeEach
    void setUp() {
        var instance = new Day12();
        setupFromFile(instance.getDefaultInputFilename());
    }

    @Test
    void sumOfAllPossibleSpringArrangements() {
        var instance = new Day12();
        int expected = 21;
        var actual = instance.sumOfAllPossibleSpringArrangements(lines, false);
        assertEquals(expected, actual, "The sum of all possible spring arrangements " +
                "is not what was expected: " + expected);
    }

    @Test
    void sumOfAllPossibleUnfoldedSpringArrangements() {
        var instance = new Day12();
        int expected = 525152;
        var actual = instance.sumOfAllPossibleSpringArrangements(lines, true);
        assertEquals(expected, actual, "he sum of all possible unfolded spring arrangements " +
                "is not what was expected: " + expected);
    }

    @Override
    protected Logger getLogger() {
        return LOGGER;
    }
}

File: aoc-2017/src/test/kotlin/com/capital7software/aoc/aoc2017aoc/days/Day12Test.kt
package com.capital7software.aoc.aoc2017aoc.days

import com.capital7software.aoc.lib.AdventOfCodeTestBase
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.slf4j.Logger
import org.slf4j.LoggerFactory

class Day12Test : AdventOfCodeTestBase() {
  companion object {
    val log: Logger = LoggerFactory.getLogger(Day12Test::class.java)
  }

  override fun getLogger(): Logger {
    return log
  }

  @BeforeEach
  fun setup() {
    val instance = Day12()
    setupFromFile(instance.defaultInputFilename)
  }

  @Test
  fun testProgramsInGroup() {
    val instance = Day12()
    val expected = 6

    val actual = instance.programsInGroup(lines, 0)

    Assertions.assertEquals(
        expected, actual, "The fewest steps " +
        "is not what was expected: $expected"
    )
  }

  @Test
  fun testAllGroups() {
    val instance = Day12()
    val expected = 2

    val actual = instance.allGroups(lines)

    Assertions.assertEquals(
        expected, actual, "The number of groups " +
        "is not what was expected: $expected"
    )
  }
}
